plugins/plugin-java/che-plugin-java-plain/che-plugin-java-plain-server/src/main/java/org/eclipse/che/plugin/java/plain/server/inject/PlainJavaProjectSourceFolderWatcher.java

```diff
@@ -0,0 +1,197 @@
+import org.eclipse.che.api.project.shared.RegisteredProject;
+import org.eclipse.che.api.watcher.server.FileWatcherManager;
+import org.eclipse.che.api.watcher.server.impl.FileWatcherByPathMatcher;
+import org.eclipse.che.commons.lang.concurrent.LoggingUncaughtExceptionHandler;
+import org.eclipse.che.plugin.java.languageserver.JavaLanguageServerExtensionService;
+import org.eclipse.lsp4j.DidChangeWatchedFilesParams;
+import org.eclipse.lsp4j.FileChangeType;
+import org.eclipse.lsp4j.FileEvent;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Reports the create/update/delete changes on project source folders to jdt.ls
+ *
+ * @author V. Rubezhny
+ */
+public class PlainJavaProjectSourceFolderWatcher {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(PlainJavaProjectSourceFolderWatcher.class);
+
+  private static final List<String> EMPTY_SOURCE_FOLDERS = emptyList();
+
+  private final FileWatcherManager manager;
+  private final FileWatcherByPathMatcher matcher;
+  private final EventService eventService;
+  private final ExecutorService executorService;
+
+  private final CopyOnWriteArrayList<Integer> watcherIds = new CopyOnWriteArrayList<>();
+  private final JavaLanguageServerExtensionService extensionService;
+  private final CopyOnWriteArrayList<LanguageServer> languageServers = new CopyOnWriteArrayList<>();
+
+  private final ProjectManager projectManager;
+
+  @Inject
+  public PlainJavaProjectSourceFolderWatcher(
+      FileWatcherManager manager,
+      FileWatcherByPathMatcher matcher,
+      EventService eventService,
+      JavaLanguageServerExtensionService extensionService,
+      ProjectManager projectManager) {
+    this.manager = manager;
+    this.matcher = matcher;
+    this.eventService = eventService;
+    this.extensionService = extensionService;
+    this.projectManager = projectManager;
+    this.executorService =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat("WorkspaceUpdater-%d")
+                .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                .setDaemon(true)
+                .build());
+  }
+
+  @PostConstruct
+  protected void startWatchers() {
+    int watcherId =
+        manager.registerByMatcher(
+            folderMatcher(),
+            s -> report(s, FileChangeType.Created),
+            s -> report(s, FileChangeType.Changed),
+            s -> report(s, FileChangeType.Deleted));
+
+    watcherIds.add(watcherId);
+    eventService.subscribe(this::onServerInitialized, LanguageServerInitializedEvent.class);
+    eventService.subscribe(this::onProjectUpdated, ProjectUpdatedEvent.class);
+  }
+
+  @PreDestroy
+  public void stopWatchers() {
+    watcherIds.stream().forEach(id -> manager.unRegisterByMatcher(id));
+  }
+
+  private void onServerInitialized(LanguageServerInitializedEvent event) {
```

Comment:
No need to collect all language servers here. We only want to notify the java language server, and we can find that one by id.

Reply:
OK

Reply:
It's done exactly the same way here:
https://github.com/eclipse/che/blob/master/wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/LanguageServerFileWatcher.java#L82

If I won't wait for LS start, the obtaining of Java LS will be failed - not sure if it's a good trying to  do anything with LS before it's started...
I don't know... Maybe I can use someow `org.eclipse.che.api.languageserver.FindServer.byId(String)` - but again - it could return a number of LS-s, as well as null/empty set of LS-s (if it is allowed t use it from `che-plugin-java-plain-server`